[LIBCLOUD-1017] Improvements to Azure ARM compute driver

palashgandhi opened the pull request:
## [LIBCLOUD-1017] Improvements to Azure ARM compute driver

### Description

The following are the proposed changes in brief:

1. Add functionality to resize a volume in Azure
2. Add functionality to update the network profile of a node
3. Add functionality to update a network interface's properties
4. Add functionality to check IP address availability
5. Create and Update some fixtures
6. Some minor fixes to docstrings

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

palashgandhi commented:
@tetron Could you please review this change? I did address your concerns around the API version to be used by simply modifying the versions of the calls that didn't work for me and left the other versions as they were.

palashgandhi commented:
@pquentin I'd appreciate it if this PR could also be included in the new release. Thanks in advance.

pquentin commented:
Sorry, I don't have the bandwith to review this myself. But if someone else does, it will be included in the new release.

palashgandhi commented:
@vdloo could you please spare some time to take a look at this PR too?

vdloo commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -1811,15 +1951,15 @@ def ex_create_tags(self, resource, tags, replace=False):
         r = self.connection.request(
             resource,
-            params={"api-version": RESOURCE_API_VERSION})
+            params={"api-version": '2018-06-01'})
```
Why hardcode the '2018-06-01' here? would this make it incompatible with the '2016-04-30-preview' resource API? Also if any api version is going to be configured it would be nicer to put it in a module level variable at the top of the file like RESOURCE_API_VERSION.

vdloo replied:
I see that there are other api-versions hardcoded in this file already as well, for example [here](https://github.com/apache/libcloud/blob/080b343e0983f2a22f7647b4fd8cebb3339adbe2/libcloud/compute/drivers/azure_arm.py#L1411), so perhaps that is not an issue.

palashgandhi replied:
@vdloo Thanks for the feedback. I had the same thought that the driver should be using the same version for each API call, but valid objections were raised on an older (now discarded) PR: https://github.com/apache/libcloud/pull/1243 which modified the driver to use a single API version.

And the answer to your question about why I hardcoded the '2018-06-01' is that I was not able to tag a network interface in the 'West US 2' region with the older version.

I discarded the PR which would ensure the driver uses a single API version because it would need to be tested (not just unit testing but also end to end testing) thoroughly. While I have already been using my change in production (only to attach/detach/modify disks and network interfaces) I do not have the resources to test all the functionality of the driver at the moment.

vdloo commented:
@palashgandhi since I don't work with Azure I don't feel qualified approve and merge this PR [as I mentioned here a while back](https://github.com/apache/libcloud/pull/1231#issuecomment-410796779). But code-wise I see nothing crazy here. 

It does look there is a missing test in libcloud/test/compute/test_azure_arm.py for the newly added resize_volume though. That might be something that you could add.

palashgandhi commented:
@vdloo I have added the unit test that I missed in the previous iteration and made another minor improvement to the `resize_volume` method. Thanks once again for the feedback, can you point me to who I can ask to approve and merge this PR?

vdloo commented:
Hi @palashgandhi that sounds good. Perhaps you can try the [Libcloud developers mailinglist](https://libcloud.apache.org/community.html) to request to get some eyes on this .

palash-gandhi committed:
[LIBCLOUD-1017] Improvements to Azure ARM compute driver

palashgandhi commented:
@vdloo I did reach out on the mailing list and it seems we do not have anyone familiar with Azure. It's been 7 months and I (and my colleagues at Delphix Corp) would really like to stop maintaining our own version of libcloud. We have been using libcloud with this change in production for the last 7 months and we have seen stable results.

Could you please merge this? I will fix any issues/breakages this change causes.

jmspring commented:
I’ll take a look and review.  I’ve reviewed a previous Azure PR.  I’m also a committee for JClouds - primarily based on azure work.

palashgandhi commented:
@jmspring Could you please take a look at this PR?

Kami commented:
@palashgandhi Since this is a bigger contribution, can you please sign an ICLA (http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes)?

Kami commented:
Overall, from a high level perspective the change looks reasonable to me (I still need to dive in and check Azure API docs).

@c-w  Do you see any potential issues with this PR or you are OK with me merging it?

Kami commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -997,6 +997,53 @@ def attach_volume(self, node, volume, ex_lun=None,
             })
         return True
 
+    def resize_volume(self, volume, new_size, ex_resource_group):
```
This method is not part of the standard (common) compute API, so it should be prefixed with ``ex_`` ( ``ex_resize_volume``).

Kami commented:
I've added a small comment about method name, besides that, LGTM.

Please submit your ICLA so we can get this merged. Thanks.

palashgandhi commented:
@Kami I have submitted the ICLA and addressed your comment. Are there a lot of queued builds in progress at the moment? The above build has been in the "Received" state for a while now.

Kami commented:
@palashgandhi Yeah, it looks like Travis has some build backlog issues. I encountered the same problem earlier.

Kami commented:
As far as ICLA goes - I will have a look (usually it takes a while for the ASF secretary to process it - it's a manual process).

Kami commented:
I made a small change (5bbdc76244b29460e46c7bac4ff9ed4e64da02d6 - Argument only needs ``ex_`` prefix if it's an extension argument which is part of a standard API method. No need for the prefix if the method itself already contains ``ex_`` prefix.) and merged changes into trunk.

Thanks again for your contribution and sorry for the delay.

palashgandhi commented:
Thank you @Kami
